ImageData/Hdf5Loader.cc

```diff
@@ -200,206 +200,172 @@ bool Hdf5Loader::UseRegionSpectralData(const std::shared_ptr<casacore::ArrayLatt
+                        case CARTA::StatsType::NanCount:
+                        case CARTA::StatsType::NumPixels:
+                            break;
+                        default:
+                            kv.second[z] = NAN;
+                            break;
+                    }
+                }
             }
         }
+    };
 
-        // get a copy of current region state
-        RegionState region_state;
-        if (!_frame->GetRegionState(region_id, region_state)) {
-            return false;
-        }
+    size_t delta_x = INIT_DELTA_CHANNEL; // since data is swizzled, third axis is x not channel
+    size_t max_x = x_start + delta_x;
+    if (max_x > num_x) {
+        max_x = num_x;
+    }
+    std::vector<float> slice_data;
 
-        // get a copy of current region configs
-        SpectralConfig config_stats;
-        if (!_frame->GetRegionSpectralConfig(region_id, config_stokes, config_stats)) {
+    for (size_t x = x_start; x < max_x; ++x) {
+        if (!GetCursorSpectralData(slice_data, stokes, x + x_min, 1, y_min, num_y, image_mutex)) {
             return false;
         }
 
-        std::map<CARTA::StatsType, std::vector<double>>* stats_values;
-        float progress;
-
-        // start the timer
-        auto t_start = std::chrono::high_resolution_clock::now();
-        auto t_latest = t_start;
-
-        // Lambda to calculate additional stats
-        auto calculate_stats = [&]() {
-            double sum_z, sum_sq_z;
-            uint64_t num_pixels_z;
-
+        for (size_t y = 0; y < num_y; y++) {
             for (size_t z = 0; z < num_z; z++) {
-                if (num_pixels[z]) {
-                    sum_z = sum[z];
-                    sum_sq_z = sum_sq[z];
-                    num_pixels_z = num_pixels[z];
-
-                    mean[z] = sum_z / num_pixels_z;
-                    rms[z] = sqrt(sum_sq_z / num_pixels_z);
-                    sigma[z] = sqrt((sum_sq_z - (sum_z * sum_z / num_pixels_z)) / (num_pixels_z - 1));
-                    if (has_flux) {
-                        flux[z] = sum_z / beam_area;
-                    }
-                } else {
-                    // if there are no valid values, set all stats to NaN except the value and NaN counts
-                    for (auto& kv : stats) {
-                        switch (kv.first) {
-                            case CARTA::StatsType::NanCount:
-                            case CARTA::StatsType::NumPixels:
-                                break;
-                            default:
-                                kv.second[z] = NAN;
-                                break;
-                        }
-                    }
+                casacore::IPosition mask_pos(mask_shape.size());
+                mask_pos(0) = x;
+                mask_pos(1) = y;
+                mask_pos(_spectral_axis) = z;
+                if (_stokes_axis >= 0) {
+                    mask_pos(_stokes_axis) = 0;
```

Comment:
In the refactor, the mask is a 3D/4D boolean array which matches the region shape.

In CASA images, pixel values are not changed for "bad" values, a boolean mask is created to indicate which pixels should be included in a calculation.  So there are two masks for a casacore region or SubImage, (1) a mask to indicate whether the pixel is included/excluded in the region, and (2) a mask to indicate whether the pixel is good/bad.

In dev, since the Region's mask is 2D, it ignores the image mask and only indicates whether the pixel is included/excluded (type 1).  In the refactor, the mask is a 3D/4D mask which applies the region mask (type 1) to the image mask (type 2) to get a 3D/4D region mask (type 2).  So the mask IPos must include the z index and stokes index.

The CartaHDF5Image mask (type 2) is set depending on the NaN pixel values.  So the dev/refactor code is equivalent in that NaN values are ignored using "if (std::isfinite(v))" in dev (which I now see should be removed in the refactor code) or the mask in the refactor.  Since the Hdf5Loader is the only user of the mask, I could revise this to use the 2D mask again and check the type 1 mask and pixel values separately as before.
 
If in the future there will be a way to create an image mask for an HDF5 image that does not change the pixel values (interactive clean?), the 2D mask will no longer be sufficient. @kswang1029 what do you think?

Reply:
@pford I think for interactive clean, we will just need to focus on CASA image only as that's the native image format of CASA. 

I think it is better for keep the flexibility of have masked in 3D/4D for HDF5 images. One use case I can think of is moment generator. Usually users can define:
1) a box in 2D so that it defines the spatial size of the output moments.
2) a range of channels to be used for moment calculations
3) a range of pixel values to be included/excluded for the moment calculations

So given the above three items, we will need a 3D/4D mask I suppose?

Reply:
Thanks; that makes sense. I think that if it's possible to do a single check using a 3D/4D mask, that's probably better than doing two checks (first 2D mask and later `isfinite`). So I'm happy to leave it like this.

Reply:
Update: Angus pointed out in our local meeting that the 3D/4D mask may be a problem if the image is very large. This may be a good reason to go back to the old behaviour here.